components/omega/src/ocn/Eos.h

```diff
@@ -261,11 +261,273 @@ class LinearEos {
       const I4 KStart = KChunk * VecLength;
       for (int KVec = 0; KVec < VecLength; ++KVec) {
          const I4 K = KStart + KVec;
+         if (K >= NVertLayers)
+            continue;
          SpecVol(ICell, K) =
              1.0_Real / (RhoT0S0 + (DRhodT * ConservTemp(ICell, K) +
                                     DRhodS * AbsSalinity(ICell, K)));
       }
    }
+
+ private:
+   I4 NVertLayers;
+};
+
+/// Functor for calculating the Brunt-Vaisala frequency using TEOS-10
+class Teos10BruntVaisalaFreq {
+ public:
+   /// Constructor for BruntVaisalaFreq
+   Teos10BruntVaisalaFreq(const HorzMesh *Mesh, const VertCoord *VCoord);
+
+   //   The functor takes the full arrays of Brunt-Vaisala frequency (inout),
+   //   the indix ICell, and the ocean tracers (conservative) temperature,
+   //   (absolute) salinity, pressure, specific volume as inputs, and outputs
+   //   the Brunt-Vaisala frequency.
+   KOKKOS_FUNCTION void operator()(Array2DReal BruntVaisalaFreq, I4 ICell,
+                                   const Array2DReal &ConservTemp,
+                                   const Array2DReal &AbsSalinity,
+                                   const Array2DReal &Pressure,
+                                   const Array2DReal &SpecVol) const {
+
+      Real PGrav = calcGrav(LatCell(ICell), ZMid(ICell, 0));
+      Real Db2Pa = 1.0e4;
+      for (int K = 0; K < NVertLayers; ++K) {
+         if (K == 0) {
+            // No Brunt-Vaisala frequency at surface
+            BruntVaisalaFreq(ICell, K) = 0.0_Real;
+         } else {
+            // Calculate Brunt-Vaisala frequency
+            Real NGrav = calcGrav(LatCell(ICell), ZMid(ICell, K));
+            Real CtInt =
+                0.5_Real * (ConservTemp(ICell, K) + ConservTemp(ICell, K - 1));
+            Real SaInt =
+                0.5_Real * (AbsSalinity(ICell, K) + AbsSalinity(ICell, K - 1));
+            Real PInt =
+                0.5_Real * (Pressure(ICell, K) + Pressure(ICell, K - 1));
+            Real SpInt = 0.5_Real * (SpecVol(ICell, K) + SpecVol(ICell, K - 1));
+            Real AlphaInt   = calcAlpha(SaInt, CtInt, PInt, SpInt);
+            Real BetaInt    = calcBeta(SaInt, CtInt, PInt, SpInt);
+            Real DSa        = AbsSalinity(ICell, K) - AbsSalinity(ICell, K - 1);
+            Real DCt        = ConservTemp(ICell, K) - ConservTemp(ICell, K - 1);
+            Real DP         = Pressure(ICell, K) - Pressure(ICell, K - 1);
+            Real GravityInt = 0.5_Real * (PGrav + NGrav);
+
+            BruntVaisalaFreq(ICell, K) = GravityInt * GravityInt *
+                                         ((1.0_Real / SpInt) / (Db2Pa * DP)) *
+                                         (BetaInt * DSa - AlphaInt * DCt);
+         }
+      }
+   }
+
+   KOKKOS_FUNCTION Real calcGrav(Real Lat, Real Z) const {
```

Comment:
do we really want to calculate latitude/depth varying gravity in here when the rest of the code uses a fixed value?

Reply:
@vanroekel This is how TEOS-10 does it. If we are ok with the error (compared to the TEOS-10 value) that will incur, than I am ok with changing this to a constant value.

Reply:
Ok, sounds good, I will change it to a constant value.